object_hasher.go

```diff
@@ -40,7 +40,15 @@ type objectHasher struct {
 }
 
 // `HashProto` returns the object hash of a given protocol buffer message.
-func (hasher *objectHasher) HashProto(pb proto.Message) ([]byte, error) {
+func (hasher *objectHasher) HashProto(pb proto.Message) (h []byte, err error) {
+	// Ensure that we can recover if the proto library panics.
+	// See: https://github.com/golang/protobuf/issues/478
+	defer func() {
```

Comment:
Swallowing panics is almost never the ideal solution. Your own issue report demonstrates that a swallowed panic by fmt.Println can itself result in an issue report that is complicated by a swallowed panic.

Reply:
This code is not swallowing the panic, but instead converting it into an error that's returned by the method to its caller.

Additionally, this [go blog post on panic and recover](https://blog.golang.org/defer-panic-and-recover) mentions that:

> The convention in the Go libraries is that even when a package uses panic internally, its external API still presents explicit error return values.

Reply:
So, the library itself is throwing the panics explicitly? Or are we catching language panics that are not actually internal to the package?

Reply:
It's not the protobuf library itself but rather a dependency of it that explicitly throws the panics. That dependency being the "reflect" package.

See: https://github.com/golang/go/blob/master/src/reflect/value.go

It's unfortunate since the correct thing to do for the reflect package should've been to return errors instead, but given that it uses panics as part of its error reporting I think that this recover is the lesser of two evils.

Reply:
To be fair, "reflect" is also a heavily used dependency in the code of this library itself, so a defensive recover is probably also warranted on those grounds alone.

Reply:
I just don’t see the merit of catching another library/package’s panics…